drivers/flash/flash_stm32_qspi.c

```diff
@@ -0,0 +1,927 @@
+		if (jesd216_bfp_erase(bfp, ti, etp) == 0) {
+			LOG_DBG("Erase %u with %02x",
+					(uint32_t)BIT(etp->exp), etp->cmd);
+		}
+		++etp;
+	}
+
+	data->page_size = jesd216_bfp_page_size(php, bfp);
+
+	LOG_DBG("Page size %u bytes", data->page_size);
+	LOG_DBG("Flash size %u bytes", flash_size);
+	return 0;
+}
+
+static int flash_stm32_qspi_init(const struct device *dev)
+{
+	const struct flash_stm32_qspi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_qspi_data *dev_data = DEV_DATA(dev);
+	struct dma_config dma_cfg = dev_data->dma.cfg;
+	static DMA_HandleTypeDef hdma;
+	uint32_t ahb_clock_freq;
+	uint32_t prescaler = 0;
+	int ret;
+
+	/* Signals configuration */
+	ret = stm32_dt_pinctrl_configure(dev_cfg->pinctrl_list,
+					 dev_cfg->pinctrl_list_size,
+					 (uint32_t)dev_cfg->regs);
+	if (ret < 0) {
+		LOG_ERR("QSPI pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	/* Clock configuration */
+	__ASSERT_NO_MSG(device_get_binding(STM32_CLOCK_CONTROL_NAME));
+
+	if (clock_control_on(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			     (clock_control_subsys_t) &dev_cfg->pclken) != 0) {
+		LOG_DBG("Could not enable QSPI clock");
+		return -EIO;
+	}
+
+	if (clock_control_get_rate(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			(clock_control_subsys_t) &dev_cfg->pclken,
+			&ahb_clock_freq) < 0) {
+		LOG_DBG("Failed to get_AHB clock frequency");
+		return -EIO;
+	}
+
+	for (; prescaler <= STM32_QSPI_CLOCK_PRESCALER_MAX; prescaler++) {
+		uint32_t clk = ahb_clock_freq / (prescaler + 1);
+
+		if (clk <= dev_cfg->max_frequency) {
+			break;
+		}
+	}
+	__ASSERT_NO_MSG(prescaler <= STM32_QSPI_CLOCK_PRESCALER_MAX);
+
+	/*
+	 * DMA configuration
+	 * Due to use of QSPI HAL API in current driver,
+	 * both HAL and Zephyr DMA drivers should be configured.
+	 * The required configuration for Zephyr DMA driver should only provide
+	 * the minimum information to inform the DMA slot will be in used and
+	 * how to route callbacks.
+	 */
+
+	if (dev_data->dma.name != NULL) {
+		dev_data->dma.dev = device_get_binding(dev_data->dma.name);
+		if (!dev_data->dma.dev) {
+			LOG_ERR("%s device not found", dev_data->dma.name);
+			return -ENODEV;
+		}
+	}
+
+	/* Proceed to the minimum Zephyr DMA driver init */
+	dma_cfg.user_data = &hdma;
+	/* HACK: This field is used to inform driver that it is overridden */
+	dma_cfg.linked_channel = 0x7F;
+	ret = dma_config(dev_data->dma.dev, dev_data->dma.channel, &dma_cfg);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* Proceed to the HAL DMA driver init */
+	if (dma_cfg.source_data_size != dma_cfg.dest_data_size) {
+		LOG_ERR("Source and destination data sizes not aligned");
+		return -EINVAL;
+	}
+
+	int index = find_lsb_set(dma_cfg.source_data_size) - 1;
+
+	/* TODO: This init is limited to G0/G4/L0/L4/L5/WB */
+	hdma.Init.PeriphDataAlignment = table_p_size[index];
+	hdma.Init.MemDataAlignment = table_m_size[index];
+	hdma.Init.Request = dma_cfg.dma_slot;
```

Comment:
I was trying to make this work for stm32f7, but get a compile error here when building `samples/drivers/flash_shell`.

If you could extend support to `DMA_STM32_V1` also I can help to verify it on _stm32f746g_disco_ and _stm32f769i_disco_ boards.
 
If not, perhaps use `#ifndef CONFIG_DMA_STM32_V1` or similar to make it compile.

In case you choose to add this, here's my attempt at dts files for stm32f746i_disco. I can also make such patch for stm32f769g_disco if you like.

`boards/arm/stm32f756g_disco/stm32f756_disco.dts`:
```
&dma2 {
	status = "okay";
};

&quadspi {
	pinctrl-0 = <&quadspi_clk_pb2 &quadspi_bk1_ncs_pb6
		     &quadspi_bk1_io0_pd11 &quadspi_bk1_io1_pd12
		     &quadspi_bk1_io2_pe2 &quadspi_bk1_io3_pd13>;
	dmas = <&dma2 7 3 0x0000 0x03>;
	dma-names = "tx_rx";

	status = "okay";

	mx25r6435f: qspi-nor-flash@0 {
		compatible = "st,stm32-qspi-nor";
		label = "N25Q128A13EF840E";
		reg = <0>;
		spi-max-frequency = <108000000>;
		/* 128 Megabits = 16 Megabytes */
		size = <0x8000000>;
		status = "okay";
	};
};
```
`dts/arm/st/f7/stm32f7.dtsi`:
```
		quadspi: quadspi@a0001000 {
			compatible = "st,stm32-qspi";
			#address-cells = <1>;
			#size-cells = <0>;
			reg = <0xa0001000 0x34>;
			interrupts = <92 0>;
			clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000002>;
			status = "disabled";
			label = "QUADSPI";
		};

```
(I can provide `.dts` files for QUADSPI support on these boards and theirs corresponding MCUs if it helps you.)

Reply:
@hjuul thanks for your interest. I'll try to add F7 support later this week.